src/util/netif.cpp

```diff
@@ -0,0 +1,305 @@
+    if (service.SetSockAddr(addr)) {
+        return (CNetAddr)service;
+    }
+    return std::nullopt;
+}
+
+//! MacOS: Get default gateway from route table. See route(4) for the format.
+static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    // net.route.0.inet[6].flags.gateway
+    int mib[] = {CTL_NET, PF_ROUTE, 0, family, NET_RT_FLAGS, RTF_GATEWAY};
+    // The size of the available data is determined by calling sysctl() with oldp=nullptr. See sysctl(3).
+    size_t l = 0;
+    if (sysctl(/*name=*/mib, /*namelen=*/sizeof(mib) / sizeof(int), /*oldp=*/nullptr, /*oldlenp=*/&l, /*newp=*/nullptr, /*newlen=*/0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl length of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    std::vector<std::byte> buf(l);
+    if (sysctl(/*name=*/mib, /*namelen=*/sizeof(mib) / sizeof(int), /*oldp=*/buf.data(), /*oldlenp=*/&l, /*newp=*/nullptr, /*newlen=*/0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl data of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    // Iterate over messages (each message is a routing table entry).
+    for (size_t msg_pos = 0; msg_pos < buf.size(); ) {
+        Assume((msg_pos + sizeof(rt_msghdr)) <= buf.size());
+        const struct rt_msghdr* rt = (const struct rt_msghdr*)(buf.data() + msg_pos);
+        const size_t next_msg_pos = msg_pos + rt->rtm_msglen;
+        Assume(rt->rtm_msglen >= sizeof(rt_msghdr) && next_msg_pos <= buf.size());
+        // Iterate over addresses within message, get destination and gateway (if present).
+        // Address data starts after header.
+        size_t sa_pos = msg_pos + sizeof(struct rt_msghdr);
+        std::optional<CNetAddr> dst, gateway;
+        for (int i = 0; i < RTAX_MAX; i++) {
+            if (rt->rtm_addrs & (1 << i)) {
+                // 2 is just sa_len + sa_family, the theoretical minimum size of a socket address.
+                Assume((sa_pos + 2) <= next_msg_pos);
+                const struct sockaddr* sa = (const struct sockaddr*)(buf.data() + sa_pos);
+                Assume(sa->sa_len >= 2 && (sa_pos + sa->sa_len) <= next_msg_pos);
+                if (i == RTAX_DST) {
+                    dst = FromSockAddr(sa);
+                } else if (i == RTAX_GATEWAY) {
+                    gateway = FromSockAddr(sa);
+                }
+                // Skip to next address.
+                sa_pos += sa->sa_len;
+            }
+        }
+        // Found default gateway?
+        if (dst && gateway && dst->IsBindAny()) { // Route to 0.0.0.0 or :: ?
+            return *gateway;
+        }
+        // Skip to next message.
+        msg_pos = next_msg_pos;
+    }
+    return std::nullopt;
+}
+
+#else
+
+// Dummy implementation.
+static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    (void)family;
+    return std::nullopt;
```

Comment:
nit, can drop the parameter name for an unused parameter and the `void` cast:
```suggestion
static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t)
{
    return std::nullopt;
```